File: Deepspace_Java/P2/src/deepspace/SpaceStation.java
/**
 * @author BOLÍVAR PELÁEZ, CLARA
 * @author RUIZ GÓMEZ, SOLEDAD
 * 2º DGIIM - PDOO - Practice 2 DeepSpace
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author sunage
 */
public class SpaceStation {
    
    private final int MAXFUEL = 100; //Shield units that are lost for each 
    //unit of firing power received.
    private final int SHIELDLOSSPERUNITSHOT = 100; //Greatest number of fuel 
    //units that a space station can have.
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    private Damage pendingDamage;
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;
    
    
    /**
    * @brief Constructor
    * @param n name
    * @param supplies a instance os SuppliesPackage
    */
    SpaceStation (String n, SuppliesPackage supplies) {
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        name = n;
        nMedals = 0;
        pendingDamage= null;
        hangar = null;
        weapons = new ArrayList<>();
        shieldBoosters = new ArrayList<>();
    }
    
    /**
    * @brief Set the amount of fuel to the value passed without ever 
    * exceeding the limit.
    * @param f value 
    */
    void assignFuelValue (float f) {
        if (f<=MAXFUEL)
            fuelUnits = f;
        else
            fuelUnits = MAXFUEL;
    }
    
    /**
    * @brief Set pending damage to null if it has no effect.
    */
    void cleanPendingDamage () {
        if(pendingDamage.hasNoEffect())
            pendingDamage = null;
    }
    
    /**
    * @brief It eliminates all mounted weapons and boosters that do not have 
    * any use left.
    */
    void cleanUpMountedItems () {
        
        Iterator<Weapon> it = weapons.iterator();
        Iterator<ShieldBooster> it2 = shieldBoosters.iterator();
        
        while(it.hasNext()){
            Weapon aux = it.next();
            if(aux.getUses() == 0)
                weapons.remove(aux);
        }
        
        while(it2.hasNext()){
            ShieldBooster aux2 = it2.next();
            if(aux2.getUses() == 0)
                shieldBoosters.remove(aux2);
        }
        
        
    }
    
    /**
    * @brief Set the reference of the hangar to null to indicate that it 
    * is not available.
    */
    void discardHangar () {
        hangar = null;
    }
    
    /**
    * @brief 
    * @param 
    */
    void discardShieldBooster (int i) {
        throw new UnsupportedOperationException();
    }
    
    /**
    * @brief If a hangar is available, it is requested to discard the booster 
    * with index i.
    * @param i index
    */
    void discardShieldBoosterInHangar (int i) {
        if(hangar!=null)
            hangar.removeShieldBooster(i);
    }
    
    /**
    * @brief Copy constructor
    * @param d another Damage instance
    */
    void discardWeapon (int i) {
        throw new UnsupportedOperationException();
    }
    
    /**
    * @brief If a hangar is available, it is requested to discard the weapon 
    * with index i.
    * @param i index
    */
    void discardWeaponInHangar (int i) {
        if(hangar!=null)
            hangar.removeWeapon(i);
        
    }
    
    /**
    * @brief Copy constructor
    * @param d another Damage instance
    */
    float fire () {
        throw new UnsupportedOperationException();
    }

    
    /**
    * @brief Getter with package visibility
    * @return ammoPower value
    */
    float getAmmoPower () {
        return (ammoPower);
    }
    
    /**
    * @brief Getter with package visibility
    * @return fuelUnits value
    */
    float getFuelUnits () {
        return (fuelUnits);
    }
    
    /**
    * @brief Getter with package visibility
    * @return Hangar instance
    */
    Hangar getHangar () {
        Hangar output = new Hangar (hangar);
        return output;
    }
    
    /**
    * @brief Getter with package visibility
    * @return name value
    */
    String getName () {
        return (name);
    }
    
    /**
    * @brief Getter with package visibility
    * @return nMedals value 
    */
    int getNMedals () {
        return(nMedals);
    }
    
    /**
    * @brief Getter with package visibility
    * @return pendingDamage instance
    */
    Damage getPendingDamage () {
        Damage output = new Damage (pendingDamage);
        return output;
    }
    
    /**
    * @brief Getter with package visibility
    * @return collection of ShieldBoosters
    */
    ArrayList<ShieldBooster> getShieldBoosters () {
        ArrayList<ShieldBooster> output = new ArrayList<>(shieldBoosters);
        return output;
    }
    
    /**
    * @brief Getter with package visibility
    * @return shieldPower instance
    */
    float getShieldPower () {
        return (shieldPower);
    }
    
    
    /**
    * @brief Getter with package visibility
    * @param d another Damage instance
    */
    float getSpeed () {
        return(fuelUnits/MAXFUEL);
    }
    
    /**
    * @brief builds a new WeaponToUI object from *this
    * @return WeaponToUI
    */
    SpaceStationToUI getUIversion () {
        return new SpaceStationToUI (this);
    }
    
    /**
    * @brief Getter with package visibility
    * @return collection of Weapons
    */
    ArrayList<Weapon> getWeapons () {
        return (weapons);
    }
    
    /**
    * @brief It tries to mount the booster with the index i inside the hangar. If 
    * a hangar is available, it removes the booster located in that position 
    * and if this operation is successful, it is added to the collection of
    * boosters in use.
    * @param i index
    */
    void mountShieldBooster (int i) {
        if (hangar != null) {
            ShieldBooster aux;
            aux = hangar.removeShieldBooster(i);
            if (aux != null)
                shieldBoosters.add(aux);
        }
    }
    
    /**
    * @brief It trys to mount the weapon with the index i inside the hangar. If 
    * a hangar is available, it removes the weapon located in that position 
    * and if this operation is successful, it is added to the collection of
    * weapons in use.
    * @param i index
    */
    void mountWeapon (int i) {
        if (hangar != null) {
            Weapon aux;
            aux = hangar.removeWeapon(i);
            if (aux != null) 
                weapons.add(aux);
        }
    }
    
    /**
    * @brief decrease in fuel units due to a displacement. The number of stored 
    * units is subtracted a fraction that is equal to the speed of the station
    * fuelUnits cannot be a negative 
    */
    void move () {
        fuelUnits -= getSpeed();
    }
    
    /**
    * @brief builds a new WeaponToUI object from *this
    * @return WeaponToUI
    */
    float protection () {
        throw new UnsupportedOperationException();
    }
    
    /**
    * @brief if *this has not hangar, param h is the new hangar. If *this already 
    * has hangar, this method does not operate
    * @param h hangar to clone
    */
    void receiveHangar (Hangar h) {
        if (hangar == null)
            hangar = new Hangar(h);
    }
    
    /**
    * @brief if *this has not hangar, returns false. Otherwise, it returns the 
    * result of trying to add the shield booster s to the hangar
    * @param s shield booster to add
    * @return true if added, false otherwise
    */
    boolean receiveShieldBooster (ShieldBooster s) {
        if (hangar == null)
            return false;
        else
            return hangar.addShieldBooster(s);
    }
    
    /**
    * @brief builds a new WeaponToUI object from *this
    * @return WeaponToUI
    */
    ShotResult receiveShot (float shot) {
        throw new UnsupportedOperationException();
    }
    
    /**
    * @brief increments ammoPower, shieldPower and fuelUnits 
    * @param s supplies package 
    */
    void receiveSupplies (SuppliesPackage s) {
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        fuelUnits += s.getFuelUnits();
    }
    
    /**
    * @brief if *this has not hangar, returns false. Otherwise, it returns the 
    * result of trying to add the weapon w to the hangar
    * @param w weapon to add
    * @return true if added, false otherwise
    */
    boolean receiveWeapon (Weapon w) {
        if (hangar == null)
            return false;
        else
            return hangar.addWeapon(w);
    }
    
    /**
    * @brief builds a new WeaponToUI object from *this
    * @return WeaponToUI
    */
    void setLoot (Loot loot) {
        throw new UnsupportedOperationException();
    }
    
    /**
    * @brief calculates the adjusted damage from param d to the weapon and shield
    * collections in *this and then saves it in pendingDamage
    * @param d Damage to adjust
    */
    void setPendingDamage (Damage d) {
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    /**
    * @brief returns true if *this is in a valid state (this implies not to have
    * any pendingDamage at all or with any effect
    * @return true if *this is in a valid state, false otherwise
    */
    boolean validState () {
        return ((pendingDamage == null) || (pendingDamage.hasNoEffect()));
    }
    
    /**
    * @brief toString
    * @return String with info about the instance
    */
    public String toString () {
        
        String output = "SpaceStation [ MAXFUEL " + MAXFUEL 
                        + " ; SHIELDLOSSPERUNITSHOT " + SHIELDLOSSPERUNITSHOT 
                        + " ; ammoPower " + ammoPower +" ; fuelUnits " + fuelUnits 
                            +" ; name " + name +" ; nMedals "+ nMedals +" ; shieldPower "+
                                shieldPower + " ; pendingDamage "+ pendingDamage.toString()+
                            " ; weapons " + weapons.toString() + " ; shieldBoosters " + 
                            shieldBoosters.toString() + " ; hangar " + hangar.toString() + " ]";
        
        return output;
    }
 
}

File: Deepspace_Ruby/P3/SpaceStation.rb
# BOLÍVAR PELÁEZ, CLARA
# RUIZ GÓMEZ, SOLEDAD
# 2º DGIIM - PDOO - Practice 3 DeepSpace

# SpaceStation
# 

module Deepspace

    class SpaceStation

        attr_reader :fuelUnits, :ammoPower, :hangar, :name, :nMedals, :pendingDamage, :shieldBoosters, :shieldPower, :speed, :weapons

        @@MAXFUEL = 100
        @@SHIELDLOSSPERUNITSHOT = 0.1
        
        # Constructores HACER
        def initialize (n, s)
            @ammoPower = s.ammoPower
            @fuelUnits = s.fuelUnits
            @shieldPower = s.shieldPower
            @name = n 
            @nMedals = 0
            @pendingDamage = nil
            @hangar = nil
            @weapons = Array.new
            @shieldBoosters = Array.new 

        end

        # Set the amount of fuel to the value passed without ever 
        # exceeding the limit.
        private
        def assignFuelValue (f)
            if (f<=@@MAXFUEL)
                @fuelUnits = f
            else
                @fuelUnits = @@MAXFUEL
            end
        end

        # Set pending damage to null if it has no effect.
        def cleanPendingDamage ()
            if (pendingDamage.hasNoEffect)
                pendingDamage = nil
            end
        end

        # It eliminates all mounted weapons and boosters that do not have 
        # any use left.
        public
        def cleanUpMountedItems ()       
           
            @weapons.each do |aux|
                if(aux.uses == 0)
                    @weapons.delete(aux)
                end
            end

            @shieldBoosters.each do |aux|
                if(aux.uses == 0)
                    @shieldBoosters.delete(aux)
                end
            end
  
        end

        # an attempt is made to discard the shield booster with index i from 
        # the current shield booster pool. In addition to losing the shield boost, 
        # it updates the pending damage if there is any
        def discardHangar ()
            @hangar = nil
        end

        def discardShieldBooster (i)
            size = @shieldBoosters.length

            if(i>=0 && i<size) 
                shieldBoosters.remove_at(i)

                if (pendingDamage != nil)
                    pendingDamage.discardShieldBooster
                    cleanPendingDamage
                end
            end
        end


        # If a hangar is available, it is requested to discard the booster 
        # with index i
        def discardShieldBoosterInHangar (i)
            if (@hangar != nil)
                @hangar.removeShieldBooster(i)
            end
        end

        # It tries to discard the weapon i of the array weapons. Apart from
        # losing the weapon, it should update the pendingDamage if there is any.
   
        def discardWeapon (i)
            size = @weapons.length

            if(i>=0 && i<size) 
                w = weapons.remove_at(i)

                if (pendingDamage != nil)
                    pendingDamage.discardWeapon(w)
                    cleanPendingDamage
                end
            end
        end

        # If a hangar is available, it is requested to discard the weapon 
        # with index i.
        def discardWeaponInHangar (i)
            if (@hangar != nil)
                @hangar.removeWeapon(i)
            end
        end

        # --
        def fire
            factor = 1

            weapons.each do |w|
                factor*=w.useIt
            end

            return ammoPower*factor

        end

        # calculates the speed of the station
         def speed
            @fuelUnits/@@MAXFUEL
        end

        # builds a new WeaponToUI object from self
        def getUIversion
            SpaceStationToUI.new(self)
        end

        # It tries to mount the booster with the index i inside the hangar. If 
        # a hangar is available, it removes the booster located in that position 
        # and if this operation is successful, it is added to the collection of
        # boosters in use.
        def mountShieldBooster (i)
            if (@hangar != nil)
                aux = @hangar.removeShieldBooster(i)
                if (aux != nil)
                    @shieldBoosters.add(aux)
                end 
            end
        end


        #It trys to mount the weapon with the index i inside the hangar. If 
        #a hangar is available, it removes the weapon located in that position 
        #and if this opera
        def mountWeapon (i)
            if(hangar != nil)
                aux = @hangar.removeWeapon(i)
                if(aux != nil)
                    @weapons.add(aux) 
                end
            end  
        end

        #Decrease in fuel units due to a displacement. The number of stored 
        #units is subtracted a fraction that is equal to the speed of the station
        #fuelUnits cannot be a negative 
        def move
            @fuelUnits -= speed
        end

        #Builds a new WeaponToUI object from *this
        def protection
            factor = 1

            shieldBoosters.each do |s|
                factor*=s.useIt
            end

            return shieldPower*factor
        end

        #If *this has not hangar, param h is the new hangar. If *this already 
        #has hangar, this method does not operate
        def receiveHangar (h)
            if(@hangar == nil)
                @hangar = Hangar.newCopy(h)
            end
        end

        #If *this has not hangar, returns false. Otherwise, it returns the 
        #result of trying to add the shield booster s to the hangar
        def receiveShieldBooster (s)
            if(@hangar == nil)
                return false
            else
                @hangar.addShieldBooster(s)
            end
        end

        
        def receiveShot (shot)
            
            myProtection=protection

            if(myProtection >= shot)
                @shieldPower-= @SHIELDLOSSPERUNITSHOT*shot 
                @shieldPower = [0.0, @shieldPower].max
                return ShotResult::RESIST 
            else
                @shieldPower=0.0
                return ShotResult::DONOTRESIST
            end
        end
        
        #Increments ammoPower, shieldPower and fuelUnits 
        def receiveSupplies (s)
            @ammoPower += s.ammoPower
            @fuelUnits+= s.fuelUnits
            @shieldPower+=s.shieldPower
        end

        #If *this has not hangar, returns false. Otherwise, it returns the 
        #result of trying to add the weapon w to the hangar
        def receiveWeapon (w)
            if(@hangar == nil)
                return false
            else
                return @hangar.addWeapon(w)
            end
        end

        def setLoot (loot)
            dealer= CardDealer.instance # o Instance

            h=loot.nHangars

            if(h>0)
                hangar = dealer.nextHangar
                receiveHangar(hangar)
            end

            elements= loot.nSupplies
            elements.times do 
                sup = dealer.nextSuppliesPackage
                receiveSupplies(sup)
            end

            elements = loot.nWeapons
            elements.times do
                weap = dealer.nextWeapon
                receiveWeapon(weap)
            end

            elements = loot.nShields
            elements.times do
                sh = dealer.nextShieldBooster
                receiveShieldBooster(sh)
            end

            medals = loot.nMedals
            @nMedals+=medals

        end

        #Calculates the adjusted damage from param d to the weapon and shield
        # collections in *this and then saves it in pendingDamage
        def setPendingDamage (d)
            @pendingDamage = d.adjust(@weapons, @shieldBoosters)
        end

        #returns true if *this is in a valid state (this implies not to have
        #any pendingDamage at all or with any effect
        def validState
            return ((@pendingDamage == nil ) || (@pendingDamage.hasNoEffect))
        end


    end
end
